agents/dqn.py

```diff
@@ -1,137 +1,132 @@
         )
 
-        # Shape: [batch_size]
-        best_next_neighbor_idx = neighbor_node_ids.padded_values[torch.argmax(next_neighbor_q, dim=-1)]
-        next_neighbor_idx = _with_random_research(
-            best_next_neighbor_idx,
-            neighbor_node_ids,
-            self._research_prob
-        )
-
         if self._bag_trajectory_memory is not None:
             # TODO[Vladimir Baikalov]: Think about how to generalize
-            bag_ids = inputs.get(self._bag_idx_prefix, None)
+            bag_ids = inputs.get(self._bag_ids_prefix, None)
 
             self._bag_trajectory_memory.add_to_trajectory(
                 bag_ids=bag_ids,
-                node_idxs=current_node_idx,
+                node_idxs=torch.full([batch_size], self._node_id),
                 extra_infos=zip(
-                    current_node_idx,
+                    torch.unsqueeze(current_node_idx, dim=1),
                     neighbor_node_ids,
-                    destination_node_idx,
-                    next_neighbor_idx
+                    neighbors_q,
+                    torch.unsqueeze(destination_node_idx, dim=1),
+                    torch.unsqueeze(next_neighbor_ids, dim=1)
                 )
             )
-        return {
-            'predicted_next_node_idx': next_neighbor_idx,
-            'predicted_next_node_q': next_neighbor_q,
-        }
+        inputs[self._output_prefix] = next_neighbor_ids
+        inputs.update({
+            'predicted_next_node_idx': next_neighbor_ids,
+            'predicted_next_node_q': neighbors_q,
+        })
+        return inputs
 
-    def learn(self):
-        for trajectory in self._bag_trajectory_memory.sample_trajectories_for_node_idx(
-                self._node_idx,
-                self._trajectory_sample_size
-        ):
+    def learn(self) -> Optional[Tensor]:
+        loss = 0
+        learn_trajectories = self._bag_trajectory_memory.sample_trajectories_for_node_idx(
+            self._node_id,
+            self._trajectory_sample_size,
+            1
+        )
+        if not learn_trajectories:
+            return None
+        for trajectory in learn_trajectories:
             target = trajectory[0]['reward']
-            current_node_idx, neighbor_node_ids, destination_node_idx, next_neighbor = trajectory[0]['extra_infos']
-            if len(trajectory) > 0:
-                current_node_idx_, neighbor_node_ids_, destination_node_idx_, _ = trajectory[1]['extra_infos']
-                neighbor_q_ = self._q_network(
-                    current_node_idx=current_node_idx_,
-                    neighbor_node_ids=neighbor_node_ids_,
-                    destination_node_idx=destination_node_idx_
-                )
-                target += self._discount_factor * neighbor_q_
-            neighbor_q = self._q_network(
+            current_node_idx, neighbor_node_ids, _, destination_node_idx, next_neighbor = trajectory[0]['extra_info']
+            if len(trajectory) > 1:
```

Comment:
It seems like you are working only with very short trajectories (of size 1 or 2). Is it intended?

Reply:
I thought that dqn takes only current state and next is exists, or it is not right?

Reply:
I re-checked the paper. You are right